common/src/main/java/io/pravega/common/concurrent/AsyncSemaphore.java

```diff
@@ -0,0 +1,243 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.concurrent;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.Exceptions;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A synchronization primitive that allows executing arbitrary (concurrent) tasks, where each task requires a well-known
+ * number of credits, subject to a total number of credits being available. Each task's successful execution will "borrow"
+ * its share of credits, and a task cannot execute if the number of credits available is insufficient. Credits may be
+ * restored externally using the {@link #release} method.
+ *
+ * This is similar to {@link java.util.concurrent.Semaphore}, except that this class allows for asynchronous processing
+ * and each task can request an arbitrary number of credits. It can be useful in solving problems making use of the
+ * Leaky Bucket Algorithm (https://en.wikipedia.org/wiki/Leaky_bucket).
+ */
+@ThreadSafe
+@Slf4j
+public class AsyncSemaphore implements AutoCloseable {
+    //region Members
+
+    private final long totalCredits;
+    private final String logId;
+    @GuardedBy("queue")
+    private long usedCredits;
+    @GuardedBy("queue")
+    private final ArrayDeque<PendingTask> queue;
+    @GuardedBy("queue")
+    private boolean closed;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the {@link AsyncSemaphore} class.
+     *
+     * @param totalCredits Total number of available credits.
+     * @param usedCredits  Initial number of used credits.
+     * @param logId        A log-friendly identifier for this {@link AsyncSemaphore}.
+     */
+    public AsyncSemaphore(long totalCredits, long usedCredits, String logId) {
```

Comment:
I feel like this should take a threadpool to run the tasks on. I seems a bit strange that they get run in the thread that calls release.